vm/src/hint_processor/builtin_hint_processor/bootloader/program_hash.rs

```diff
@@ -0,0 +1,209 @@
+    NoEntrypoint,
+
+    #[error(
+        "Invalid program builtin: builtin name too long to be converted to field element: {0}"
+    )]
+    InvalidProgramBuiltin(&'static str),
+
+    #[error("Invalid program data: data contains relocatable(s)")]
+    InvalidProgramData,
+
+    /// Conversion from Felt252 to FieldElement failed. This is unlikely to happen
+    /// unless the implementation of Felt252 changes and this code is not updated properly.
+    #[error("Conversion from Felt252 to FieldElement failed")]
+    Felt252ToFieldElementConversionFailed,
+}
+
+/// Computes a hash chain over the data, in the following order:
+///     h(data[0], h(data[1], h(..., h(data[n-2], data[n-1])))).
+///
+/// Reimplements this Python function:
+/// def compute_hash_chain(data, hash_func=pedersen_hash):
+///     assert len(data) >= 1, f"len(data) for hash chain computation must be >= 1; got: {len(data)}."
+///     return functools.reduce(lambda x, y: hash_func(y, x), data[::-1])
+fn compute_hash_chain<'a, I>(
+    data: I,
+    hash_func: HashFunction,
+) -> Result<FieldElement, HashChainError>
+where
+    I: Iterator<Item = &'a FieldElement> + DoubleEndedIterator,
+{
+    match data.copied().rev().reduce(|x, y| hash_func(&y, &x)) {
+        Some(result) => Ok(result),
+        None => Err(HashChainError::EmptyData),
+    }
+}
+
+/// Creates an instance of `FieldElement` from a builtin name.
+///
+/// Converts the builtin name to bytes then attempts to create a field element from
+/// these bytes. This function will fail if the builtin name is over 31 characters.
+fn builtin_to_field_element(builtin: &BuiltinName) -> Result<FieldElement, ProgramHashError> {
+    // The Python implementation uses the builtin name without suffix
+    let builtin_name = builtin
+        .name()
+        .strip_suffix("_builtin")
+        .unwrap_or(builtin.name());
+
+    FieldElement::from_byte_slice_be(builtin_name.as_bytes())
+        .map_err(|_| ProgramHashError::InvalidProgramBuiltin(builtin.name()))
+}
+
+/// The `value: FieldElement` is `pub(crate)` and there is no accessor.
+/// This function converts a `Felt252` to a `FieldElement` using a safe, albeit inefficient,
+/// method.
+fn felt_to_field_element(felt: &Felt252) -> Result<FieldElement, ProgramHashError> {
+    let bytes = felt.to_be_bytes();
+    FieldElement::from_bytes_be(&bytes)
+        .map_err(|_e| ProgramHashError::Felt252ToFieldElementConversionFailed)
+}
+
+/// Converts a `MaybeRelocatable` into a `FieldElement` value.
+///
+/// Returns `InvalidProgramData` if `maybe_relocatable` is not an integer
+fn maybe_relocatable_to_field_element(
+    maybe_relocatable: &MaybeRelocatable,
+) -> Result<FieldElement, ProgramHashError> {
+    let felt = maybe_relocatable
+        .get_int_ref()
+        .ok_or(ProgramHashError::InvalidProgramData)?;
+    felt_to_field_element(felt)
+}
+
+#[allow(dead_code)] // TODO: remove
+/// Computes the hash of a program.
```

Comment:
```suggestion
/// Computes the Pedersen hash of a program.
```

Or some other hint about what the hash impl is